native/dgehrd.go

```diff
@@ -0,0 +1,178 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack"
+)
+
+// Dgehrd reduces a block of a real n×n general matrix A to upper Hessenberg
+// form H by an orthogonal similarity transformation Q^T * A * Q = H.
+//
+// The matrix Q is represented as a product of (ihi-ilo) elementary
+// reflectors
+//  Q = H_{ilo} H_{ilo+1} ... H_{ihi-1}.
+// Each H_i has the form
+//  H_i = I - tau[i] * v * v^T
+// where v is a real vector with v[0:i+1] = 0, v[i+1] = 1 and v[ihi+1:n] = 0.
+// v[i+2:ihi+1] is stored on exit in A[i+2:ihi+1,i].
+//
+// On entry, a contains the n×n general matrix to be reduced. On return, the
+// upper triangle and the first subdiagonal of A will be overwritten with the
+// upper Hessenberg matrix H, and the elements below the first subdiagonal, with
+// the slice tau, represent the orthogonal matrix Q as a product of elementary
+// reflectors.
+//
+// The contents of a are illustrated by the following example, with n = 7, ilo =
+// 1 and ihi = 5.
+// On entry,
+//  [ a   a   a   a   a   a   a ]
+//  [     a   a   a   a   a   a ]
+//  [     a   a   a   a   a   a ]
+//  [     a   a   a   a   a   a ]
+//  [     a   a   a   a   a   a ]
+//  [     a   a   a   a   a   a ]
+//  [                         a ]
+// on return,
+//  [ a   a   h   h   h   h   a ]
+//  [     a   h   h   h   h   a ]
+//  [     h   h   h   h   h   h ]
+//  [     v1  h   h   h   h   h ]
+//  [     v1  v2  h   h   h   h ]
+//  [     v1  v2  v3  h   h   h ]
+//  [                         a ]
+// where a denotes an element of the original matrix A, h denotes a
+// modified element of the upper Hessenberg matrix H, and vi denotes an
+// element of the vector defining H_i.
+//
+// ilo and ihi determine the block of A that will be reduced to upper Hessenberg
+// form. It must hold that 0 <= ilo <= ihi <= max(0, n-1), otherwise Dgehrd will
```

Comment:
Here is the place with `ilo` and `ihi` when `n == 0` I was refering to before, @kortschak .